lib/publish.js

```diff
@@ -137,7 +137,12 @@ class Publish extends BaseCommand {
   publishConfigToOpts (publishConfig) {
     // create a new object that inherits from the config stack
     // then squash the css-case into camelCase opts, like we do
-    return flatten({...this.npm.config.list[0], ...publishConfig})
+    const mergedConfigs = Object.assign(
+      {},
+      ...[...this.npm.config.list].reverse(),
+      publishConfig
+    )
```

Comment:
ok this seem to work, I need to reverse the config.list first when merging them

Reply:
Ah ok I think I see what happened here. Your tests probably started failing cause their config.list arrays don't reflect what the real config does.  In the real config `npm.config.list[0]` has every value from all the higher-indexed configs.  So you only need to worry about the first one in the list.  My original suggestion of object spreading the results of flatten on both publishConfig and this.npm.config.list[0] is the right solution I think.

Reply:
hmm the other tests failed too 🤔  https://github.com/npm/cli/runs/2117657010?check_suite_focus=true#step:5:94

Reply:
Ok figured out the test failures.  It's because `flatten` blows up if you pass it just the publishConfig, since it is not a full config item.
This makes the tests pass in `latest` and I would assume also make them pass in your branch.

```
return flatten({...flatten(this.npm.config.list[0]), ...publishConfig})
```

Reply:
It's goofy that we have to double-flatten but that's the shortest step between here and any long term fix we implement.

Reply:
I see I see, I've committed this but found out that the return value of `flatten` isn't that compatible as an argument for another `flatten`

This test case of mine failed:

```
const tag = 'better-tag'

const configList = [
    { ...defaults, tag },
    { ...defaults, registry: `https://other.registry`, tag: 'some-tag' },
    defaults,
  ]

const Publish = requireInject('../../lib/publish.js', {
    libnpmpublish: {
      publish: (manifest, tarData, opts) => {
        t.same(opts.defaultTag, tag, 'gets option for expected tag') // defaultTag here is undefined!?
      },
    },
  })
```

This is because `flatten(this.npm.config.list[0])` takes in an object that has `tag` field and will return an object that has a field named `defaultTag`; and on the second `flatten` call, this field will never be read, i.e. `flatten(flatten(this.npm.config.list[0]))`, which causes `defaultTag` field to be undefined